tripal/tripal
# Syncing of Pubs in Tv3 Legacy Mode

Issue #577 

## Type(s) of Change(s)


## Description


The original intent for this was to fix the problem described in issue #577.  The problem occurs when Tv3 is run in legacy mode, all nodes are converted to entities but the site is still using the Tv2 templates and therefore the legacy modules are still enabled.  When importing new publications Tripal detects that the tripal_pub legacy module is enabled and  automatically tries to sync new publications as nodes.   It should not do that.    

This PR fixes the problem in the following way:
1.  When importing publications via the drush command, they are no longer automatically "syncing". Now there are new arguments that allow the user to sync (nodes) or publish (entities) . 
2.  If the user clicks the "import" link on the Chado Pub Bulk Importer page then it will automatically publish as entities (never as nodes). 

Also, this issue was a bit of a rabbit hole for the following reasons:
1.  There was duplicated code for importing pubs in multiple functions. I consolidated and cleaned up this code.
2.  The job logging wasn't working for importing pubs.  So, I fixed that, but  if importing pubs is run via a Drush command I had to duplicate log messages for the terminal.  To cut down on the amount of code I added a new argument to the tripal_report_error() function that let me pass in a Tripaljob object and that function would log to the terminal and to the job if one was available.
3.  For some reason the tripal_pub and tripal_contact modules still had some function calls in the install files that were moved over to the tripal_chado module and when I enabled them I had problems.  When testing my fixes enabling these modules failed.   I am not certain why non one who is running in legacy mode didn't encounter these problems... Maybe they did and resolved them... don't know. But I fixed it.

## Testing?





Perform steps 1-3 prior to pulling the branch for this PR.

1. To test this you will need to enable the tripal_pub legacy module (will enable all dependencies as well).
2.  Next go to Tripal > Data Loaders > Chado Publication Bulk Importer and create an importer if you don't already have one.  I found that an easy one to test with few publications is to set the number of days to 30, and the search term to 'Citrus'.  It returns just a few publications.  If you have some importers already you might want to deactivate them temporarily to keep this test from taking too long.
3.  Now try to run the importer using drush.  Notice it will try to sync the publications.
```
  drush trp-import-pubs --username=[username] --update=Y
``` 
4.  Pull the branch for this PR.
5.  Repeat the drush command from step 3.   It should now try to publish instead of syncing.
6.  Now run this drush command. It should try to sync but not publish
```
  drush trp-import-pubs --username=[username] --update=Y --sync=Y --publish=N
```
7.  Now run this drush command. It should try to both sync and publish.
```
  drush trp-import-pubs --username=[username] --update=Y --sync=Y --publish=Y
```
8.  Now run the following drush command.  It will neither try to sync or publish.
```
  drush trp-import-pubs --username=[username] --update=Y --sync=N --publish=N
```
9.  Now run this drush command and play with the various combinations of --sync and --publish to check all combinations like steps 6-8 above.
```
  drush trp-import-pubs --username=[username] --dbxref=PMID:24163125  --update=Y --sync=N --publish=Y
```
10.  Now go to the Tripal > Data Loaders > Chado Publication Bulk Importer and click the link to  "Import Pubs" next to the importer (choose one that is not disabled and that runs quickly).  It will submit a job that you can run with the provided drush command.  It should try to publish.   
11.  Go to the Tripal > Jobs page and click on the "Import Publications" link for the job you just ran.  You should see a page with the job details and the job log.   Open the Job logs and make sure you see the logs.

## Screenshots (if appropriate):

## Additional Notes (if any):
